include/nbl/builtin/hlsl/tgmath.hlsl

```diff
@@ -4,70 +4,130 @@
-//
-//	AsUint tmp = bit_cast<AsUint>(val);
-//	return (tmp & (~ieee754::traits<AsFloat>::signMask)) == ieee754::traits<AsFloat>::inf;
-//}
-//
-//#ifdef __HLSL_VERSION
-//#define DEFINE_IS_NAN_SPECIALIZATION(TYPE)\
-//template<>\
-//inline bool isNaN<TYPE>(TYPE val)\
-//{\
-//	return spirv::isNan(val);\
-//}\
-//
-//#define DEFINE_IS_INF_SPECIALIZATION(TYPE)\
-//template<>\
-//inline bool isInf<TYPE>(TYPE val)\
-//{\
-//	return spirv::isInf(val);\
-//}\
-//
-//DEFINE_IS_NAN_SPECIALIZATION(float16_t)
-//DEFINE_IS_NAN_SPECIALIZATION(float32_t)
-//DEFINE_IS_NAN_SPECIALIZATION(float64_t)
-//
-//DEFINE_IS_INF_SPECIALIZATION(float16_t)
-//DEFINE_IS_INF_SPECIALIZATION(float32_t)
-//DEFINE_IS_INF_SPECIALIZATION(float64_t)
-//
-//#undef DEFINE_IS_INF_SPECIALIZATION
-//#undef DEFINE_IS_NAN_SPECIALIZATION
-//#undef INTRINSIC_FUNC_NAMESPACE
-//#endif
-
-}
-
-}
-}
-
-#endif
\ No newline at end of file
+// TODO: will not work for emulated_float as an input because `concepts::floating_point<T>` is only for native floats, fix every occurance
+template<typename FloatingPoint NBL_FUNC_REQUIRES(concepts::FloatingPointLikeScalar<FloatingPoint>)
+inline FloatingPoint erf(FloatingPoint x)
+{
+    return tgmath_impl::erf_helper<FloatingPoint>::__call(x);
+}
+
+template<typename FloatingPoint NBL_FUNC_REQUIRES(concepts::FloatingPointLikeScalar<FloatingPoint>)
+inline FloatingPoint erfInv(FloatingPoint x)
+{
+    return tgmath_impl::erfInv_helper<FloatingPoint>::__call(x);
+}
+
+template<typename T NBL_FUNC_REQUIRES(concepts::FloatingPointLikeScalar<T> || concepts::Vectorial<T>)
+inline T floor(NBL_CONST_REF_ARG(T) val)
+{
+    return tgmath_impl::floor_helper<T>::__call(val);
+}
+
+template<typename T, typename U NBL_FUNC_REQUIRES((concepts::FloatingPointLikeScalar<T> || concepts::FloatingPointLikeVectorial<T>) && (concepts::FloatingPointLikeScalar<U> || is_same_v<U, bool>))
+inline T lerp(NBL_CONST_REF_ARG(T) x, NBL_CONST_REF_ARG(T) y, NBL_CONST_REF_ARG(U) a)
+{
+    return tgmath_impl::lerp_helper<T, U>::__call(x, y, a);
+}
+
+template<typename FloatingPoint NBL_FUNC_REQUIRES(concepts::FloatingPointLikeScalar<FloatingPoint>)
+inline bool isnan(NBL_CONST_REF_ARG(FloatingPoint) val)
+{
+    return tgmath_impl::isnan_helper<FloatingPoint>::__call(val);
+}
+
+template<typename V NBL_FUNC_REQUIRES(concepts::FloatingPointLikeVectorial<V>)
+inline vector<bool, vector_traits<V>::Dimension> isnan(NBL_CONST_REF_ARG(V) val)
+{
+    return tgmath_impl::isnan_helper<V>::__call(val);
+}
+
+template<typename FloatingPoint NBL_FUNC_REQUIRES(concepts::FloatingPointLikeScalar<FloatingPoint>)
+inline FloatingPoint isinf(NBL_CONST_REF_ARG(FloatingPoint) val)
+{
+    return tgmath_impl::isinf_helper<FloatingPoint>::__call(val);
+}
+
+template<typename V NBL_FUNC_REQUIRES(concepts::FloatingPointLikeVectorial<V>)
+inline vector<bool, vector_traits<V>::Dimension> isinf(NBL_CONST_REF_ARG(V) val)
+{
+    return tgmath_impl::isinf_helper<V>::__call(val);
+}
```

Comment:
instead of overload, maybe you just want to require that this `_helper` declares a `using return_t = ` ?

Reply:
leaving as it is

Reply:
actually i will do it

Reply:
done here: https://github.com/Devsh-Graphics-Programming/Nabla/pull/817